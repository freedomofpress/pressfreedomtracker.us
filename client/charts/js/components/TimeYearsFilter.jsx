import React from 'react'
import { countBy, range } from 'lodash'
import BarChartYears from './BarChartYears'
import CheckBoxesYear from './CheckBoxesYear'

export default function TimeYearsFilter({
	width,
	height,
	dateExtents,
	dataset,
	filterParameters: selectedYears,
	setFilterParameters: setSelectedYears,
}) {
	const [minDateYear, maxDateYear] = dateExtents.map((date) => date.getFullYear())
	const allYears = range(minDateYear, maxDateYear)

	const initializeYears = Object.fromEntries(allYears.map((year) => [year, 0]))

	const countedYears = countBy(dataset, (d) => {
		return new Date(d.date).getFullYear()
	})

	const years = { ...initializeYears, ...countedYears }

	const countYears = Object.entries(years).map(([year, count]) => ({
		year: Number(year),
		count,
	}))

	function onYearClick(d) {
		const newYears = (oldYears) =>
			oldYears.includes(d.year) ? oldYears.filter((year) => year !== d.year) : [...oldYears, d.year]
		setSelectedYears('filterTimeYears', newYears)
	}

	return (
		<div style={{ flexDirection: 'row' }}>
			<BarChartYears
				width={width}
				height={height}
				countYears={countYears}
				selectedYears={selectedYears}
				setSelectedYears={setSelectedYears}
				onClick={onYearClick}
			/>

			<CheckBoxesYear
				width={width}
				options={countYears}
				selectedYears={selectedYears}
				setSelectedYears={setSelectedYears}
				onClick={onYearClick}
			/>
		</div>
	)
}
