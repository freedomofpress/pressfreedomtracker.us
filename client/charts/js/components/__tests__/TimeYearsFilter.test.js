import React from 'react'
import * as d3 from 'd3'
import ShallowRenderer from 'react-test-renderer/shallow'
import TimeYearsFilter from '../TimeYearsFilter'

test('renders TimeYearsFilter with mocked data', () => {
	const renderer = new ShallowRenderer();
	expect(renderer.render(
		<TimeYearsFilter
			width={500}
			height={500}
			dateExtents={[d3.timeParse('%m-%d-%Y')('1-1-2020'), d3.timeParse('%m-%d-%Y')('1-1-2022')]}
			dataset={[{ date: d3.timeParse('%m-%d-%Y')('6-1-2020') }]}
			filterParameters={['2020']}
		/>
	)).toMatchSnapshot();
});
