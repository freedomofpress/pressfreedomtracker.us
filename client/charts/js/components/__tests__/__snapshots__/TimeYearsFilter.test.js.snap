// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TimeYearsFilter with mocked data 1`] = `
<div
  style={
    {
      "flexDirection": "row",
    }
  }
>
  <BarChartYears
    countYears={
      [
        {
          "count": 1,
          "year": 2020,
        },
        {
          "count": 0,
          "year": 2021,
        },
      ]
    }
    height={500}
    onClick={[Function]}
    selectedYears={
      [
        "2020",
      ]
    }
    width={500}
  />
  <CheckBoxesYear
    onClick={[Function]}
    options={
      [
        {
          "count": 1,
          "year": 2020,
        },
        {
          "count": 0,
          "year": 2021,
        },
      ]
    }
    selectedYears={
      [
        "2020",
      ]
    }
    width={500}
  />
</div>
`;
